ceph.spec.in

```diff
@@ -225,10 +226,10 @@ BuildRequires:	patch
 BuildRequires:	perl
 BuildRequires:	pkgconfig
 BuildRequires:  procps
-BuildRequires:	python%{python3_pkgversion}
-BuildRequires:	python%{python3_pkgversion}-devel
-BuildRequires:	python%{python3_pkgversion}-setuptools
-BuildRequires:	python%{python3_pkgversion}-Cython
+BuildRequires:	python%{python3_version_nodots}
```

Comment:
i don't understand this change. and i cannot find the explanation in the commit message or in tracker ticket.

Reply:
@tchaikov sure, I'll update the PR to comment my changes. In this very case, the `python3_pkgversion` macro is not used in the RPM ecosystem, while the [`python3_version_nodots` is a very well established macro](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/).

Reply:
i was looking at https://fedoraproject.org/wiki/PackagingDrafts:Python3EPEL and was referencing lots of actively maintained python packages, for instance

- https://src.fedoraproject.org/rpms/python-requests/blob/rawhide/f/python-requests.spec
- https://src.fedoraproject.org/rpms/python-setuptools/blob/rawhide/f/python-setuptools.spec

so i am not convinced that `python3_pkgversion` is not used n the RPM ecosystem.

Reply:
and this change should be put into a separate commit.

Reply:
Thanks for the reference @tchaikov . My main concern with the `python3_pkgversion` is that the name is not very self-describing, ulinke `python3_version`, `python3_version_nodots` or `python3_version_major` (which follows the [CMake convention](https://cmake.org/cmake/help/v3.14/module/FindPython3.html)).

Why should this be in a separate commit?

Reply:
> Thanks for the reference @tchaikov . My main concern with the `python3_pkgversion` is that the name is not very self-describing, ulinke `python3_version`, `python3_version_nodots` or `python3_version_major` (which follows the [CMake convention](https://cmake.org/cmake/help/v3.14/module/FindPython3.html)).
> 

IMHO, pkgversion is a good name in this context, in which it specifies the python version of a python package.

i don't understand why we need to follow the CMake convention here.  FindPython3 module follows the convention of [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection), so `find_package(Python3 ...)` can work as expected.

> Why should this be in a separate commit?

because i think it is orthogonal to "use Python 3.8".

Reply:
> IMHO, pkgversion is a good name in this context, in which it specifies the python version of a python package.

@tchaikov my main concern with that name is that it doesn't specify whether that's the major number only (as it was so far), the complete version with dots or without (as it should be, at least for CentOS and OpenSUSE `python38` and other ABI-bound packages).

> because i think it is orthogonal to "use Python 3.8".
Do you mean replacing the `python3_pkgversion` with `python3_version_nodots`?

I don't think so: maybe it's not compulsory, but you still need 2 macros to cope with packages bound to the 3.8 ABI (e.g.: `python38-Cython`) and pure source code packages (e.g.: `python3-cherrypy`). If you prefer to keep `python3_pkgversion`, which one should be it? The major-version one or the full nodots? That's why I preferred having a self-descriptive macro. What's your preference?

Reply:
i don't think we should not differentiate the python packages which contains binaries from the ones which are implemtented in "pure python". as  both of them are packaged and installed for a certain python x.y version, instead of a certain major version. take `python3-cherrypy` as an example
```console
$ rpm -q --provides python3-cherrypy
python3-cherrypy = 18.4.0-1.el8
python3.6dist(cherrypy) = 18.4.0
python3dist(cherrypy) = 18.4.0
```

the reasons why i think `python3_pkgversion` is the right choice is not completely based on by my personal preferences or taste:

1. to follow related package guide line: https://fedoraproject.org/wiki/PackagingDrafts:Python3EPEL
  
   > All references to python3 (except in macro names) need to be replaced with python%{python3_pkgversion}.

2. to follow examples. as it is used by lots of actively maintained python packages. see 
   1. https://src.fedoraproject.org/rpms/python-setuptools/blob/rawhide/f/python-setuptools.spec
   2. https://src.fedoraproject.org/rpms/python-requests/blob/rawhide/f/python-requests.spec

in general, python package maintainers should not hardwire a python package to a certain python3 minor version. python3 means **the** python3.6 in EL8, and it is python3.9 in fedora33. `python3_pkgversion` is used to encapsulate this information, and i **think** it serves this purpose well. i am not an expert on the rpm packaging, and i don't feel comfortable commenting on if other macros are superior or better than `python3_pkgversion` . it is but my 2 cents.

> Do you mean replacing the `python3_pkgversion` with `python3_version_nodots`?

yes. and i don't recommend `python3_version_nodots`. but if you insist, i'd suggest do this in two steps:

1. to replace the `python3_pkgversion` with `python3_version_nodots`. and accompany it with explanations. so far the commit message still does not explain why we need this change. and your explanations in the comments have been changing.
2. to set the default values of the rpm macros from "3" to "3.8". please note, this does not change the dependencies from 3 to 3.8, as the behavior depends on the existence of `python-srpm-macros`, `python-rpm-macros`, `python36-rpm-macros` and `python38-rpm-macros` .

instead i'd suggest:

1. install `python38-rpm-macros` in `install-deps.sh`. with which,  we have:

  ```console
  $ rpm --eval '%{python3_pkgversion}'
  38
  ```
2. replace `python3-devel` with `python38-devel` in `install-deps.sh`
3. update https://github.com/ceph/ceph-build/blob/master/scripts/build_utils.sh#L1396
4. make sure readthedoc build does not break. see https://readthedocs.org/projects/ceph/builds/14110525/ for an example, where python3.7 instead of python3.8 is used, but `typing.Literal` was not available in python3.7.

Reply:
RtD change is created at #42577